### ISSUE TITLE
_ExtInt in block parameter list causes Clang to crash

### ISSUE THREAD

**Issue by alexrp** [llvm/llvm-project#50503]
|  |  |
| --- | --- |
| Bugzilla Link | [51159](https://llvm.org/bz51159) |
| Version | 12.0 |
| OS | All |
| CC | @DougGregor,@zygoloid |

## Extended Description 
Code:

    #include <stdlib.h>
    void * _NSConcreteGlobalBlock[32];
    void foo(void)
    {
        __auto_type blk = ^void *(unsigned _ExtInt(64) len)
        {
            return malloc(len);
        };
    }

Compile with `clang test.c -fblocks -shared` or something to that effect.

Result:

    clang version 12.0.1
    Target: x86_64-pc-windows-msvc
    Thread model: posix
    InstalledDir: C:\Program Files\LLVM\bin
    "C:\\Program Files\\LLVM\\bin\\clang.exe" -cc1 -triple x86_64-pc-windows-msvc19.29.30040 -emit-obj -mrelax-all -mincremental-linker-compatible --mrelax-relocations -disable-free -disable-llvm-verifier -discard-value-names -main-file-name foo.c -mrelocation-model pic -pic-level 2 -mframe-pointer=none -fmath-errno -fno-rounding-math -mconstructor-aliases -munwind-tables -target-cpu x86-64 -tune-cpu generic -v -resource-dir "C:\\Program Files\\LLVM\\lib\\clang\\12.0.1" -internal-isystem "C:\\Program Files\\LLVM\\lib\\clang\\12.0.1\\include" -internal-isystem "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.29.30037\\include" -internal-isystem "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.29.30037\\atlmfc\\include" -internal-isystem "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.19041.0\\ucrt" -internal-isystem "C:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\shared" -internal-isystem "C:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\um" -internal-isystem "C:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\winrt" -fdebug-compilation-dir "C:\\Users\\alex\\source\\repos\\tests\\c" -ferror-limit 19 -fmessage-length=150 -fblocks -fno-use-cxa-atexit -fms-extensions -fms-compatibility -fms-compatibility-version=19.29.30040 -fdelayed-template-parsing -fcolor-diagnostics -faddrsig -o "C:\\Users\\alex\\AppData\\Local\\Temp\\foo-7c8ff7.o" -x c foo.c
    clang -cc1 version 12.0.1 based upon LLVM 12.0.1 default target x86_64-pc-windows-msvc
    #include "..." search starts here:
    #include <...> search starts here:
    C:\Program Files\LLVM\lib\clang\12.0.1\include
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\atlmfc\include
    C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\ucrt
    C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared
    C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um
    C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt
    End of search list.
    PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
    Stack dump:
    0.      Program arguments: "C:\\Program Files\\LLVM\\bin\\clang.exe" -cc1 -triple x86_64-pc-windows-msvc19.29.30040 -emit-obj -mrelax-all -mincremental-linker-compatible --mrelax-relocations -disable-free -disable-llvm-verifier -discard-value-names -main-file-name foo.c -mrelocation-model pic -pic-level 2 -mframe-pointer=none -fmath-errno -fno-rounding-math -mconstructor-aliases -munwind-tables -target-cpu x86-64 -tune-cpu generic -v -resource-dir "C:\\Program Files\\LLVM\\lib\\clang\\12.0.1" -internal-isystem "C:\\Program Files\\LLVM\\lib\\clang\\12.0.1\\include" -internal-isystem "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.29.30037\\include" -internal-isystem "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.29.30037\\atlmfc\\include" -internal-isystem "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.19041.0\\ucrt" -internal-isystem "C:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\shared" -internal-isystem "C:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\um" -internal-isystem "C:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\winrt" -fdebug-compilation-dir C:\\Users\\alex\\source\\repos\\tests\\c -ferror-limit 19 -fmessage-length=150 -fblocks -fno-use-cxa-atexit -fms-extensions -fms-compatibility -fms-compatibility-version=19.29.30040 -fdelayed-template-parsing -fcolor-diagnostics -faddrsig -o C:\\Users\\alex\\AppData\\Local\\Temp\\foo-7c8ff7.o -x c foo.c
    1.      <eof> parser at end of file
    2.      foo.c:3:6: LLVM IR generation of declaration 'foo'
    3.      foo.c:3:6: Generating code for declaration 'foo'
    #&#8203;0 0x00007ff6b9d9528a C:\Program Files\LLVM\bin\clang.exe 0x37c528a C:\Program Files\LLVM\bin\clang.exe 0x37c2e2b
    #&#8203;1 0x00007ff6b9d9528a C:\Program Files\LLVM\bin\clang.exe 0x1fecbc9 C:\Program Files\LLVM\bin\clang.exe 0x1feb5b8
    #&#8203;2 0x00007ff6b9d9528a C:\Program Files\LLVM\bin\clang.exe 0x1fe93e6 C:\Program Files\LLVM\bin\clang.exe 0x1fe799c
    #&#8203;3 0x00007ff6b9d9528a C:\Program Files\LLVM\bin\clang.exe 0x20269e1 C:\Program Files\LLVM\bin\clang.exe 0x1f1c465
    #&#8203;4 0x00007ff6b9d9528a C:\Program Files\LLVM\bin\clang.exe 0x1f21641 C:\Program Files\LLVM\bin\clang.exe 0x1f1edeb
    #&#8203;5 0x00007ff6b9d9528a C:\Program Files\LLVM\bin\clang.exe 0x1f1a7b1 C:\Program Files\LLVM\bin\clang.exe 0x1f1a334
    #&#8203;6 0x00007ff6b9d9528a C:\Program Files\LLVM\bin\clang.exe 0x1fd2acc C:\Program Files\LLVM\bin\clang.exe 0x1fc982c
    #&#8203;7 0x00007ff6b9d9528a C:\Program Files\LLVM\bin\clang.exe 0x1fc8c10 C:\Program Files\LLVM\bin\clang.exe 0x1fd3400
    #&#8203;8 0x00007ff6b9d9528a C:\Program Files\LLVM\bin\clang.exe 0x1e8bb27 C:\Program Files\LLVM\bin\clang.exe 0x1e8c3b6
    #&#8203;9 0x00007ff6b9d9528a C:\Program Files\LLVM\bin\clang.exe 0x1d156cf C:\Program Files\LLVM\bin\clang.exe 0x1d0e54c
    #&#8203;10 0x00007ff6b9d9528a C:\Program Files\LLVM\bin\clang.exe 0x1d12689 C:\Program Files\LLVM\bin\clang.exe 0x1d19022
    #&#8203;11 0x00007ff6b9d9528a C:\Program Files\LLVM\bin\clang.exe 0x3d7278f C:\Program Files\LLVM\bin\clang.exe 0x3d70489
    #&#8203;12 0x00007ff6b9d9528a C:\Program Files\LLVM\bin\clang.exe 0x2f236c9 C:\Program Files\LLVM\bin\clang.exe 0x228b562
    #&#8203;13 0x00007ff6b9d9528a C:\Program Files\LLVM\bin\clang.exe 0x225250c C:\Program Files\LLVM\bin\clang.exe 0x22f2a2e
    #&#8203;14 0x00007ff6b9d9528a C:\Program Files\LLVM\bin\clang.exe 0x7164 C:\Program Files\LLVM\bin\clang.exe 0x4344
    #&#8203;15 0x00007ff6b9d9528a C:\Program Files\LLVM\bin\clang.exe 0x4066 C:\Program Files\LLVM\bin\clang.exe 0x3d8e9b8
    #&#8203;16 0x00007ff6b9d9528a (C:\Program Files\LLVM\bin\clang.exe+0x37c528a)
    #&#8203;17 0x00007ff6b9d92e2b (C:\Program Files\LLVM\bin\clang.exe+0x37c2e2b)
    0x00007FF6B9D9528A (0x0000025C76D71108 0x0000000000000020 0x0000025C76538DA0 0x0000025C765EEB70)
    0x00007FF6B9D92E2B (0x0000000000000000 0x0000000000000000 0x0000000000000000 0x0000000000000000)
    0x00007FF6B85BCBC9 (0x0000025C76E40128 0x0000000000000000 0x0000025C76E40128 0x0000B7E40B3B486E)
    0x00007FF6B85BB5B8 (0x00007FF6BC017940 0x000000AB48588558 0x000000AB48588580 0x000000AB48587F60)
    0x00007FF6B85B93E6 (0x0000025C765EEA10 0x00007FFB135D00A6 0x0000000000000000 0x0000000000000400)
    0x00007FF6B85B799C (0xFFFFFFFFFFFFF000 0x0000025C76E400F0 0x0000025C765EEA50 0x0000B7E40B3BA6DE)
    0x00007FF6B85F69E1 (0x0000000000000008 0x0000025C00000008 0x0000B7E40B3BA6AE 0x0000000000000040)
    0x00007FF6B84EC465 (0x0000000000000000 0x0000000000000000 0x0000025C76E400F0 0x0000000000000080)
    0x00007FF6B84F1641 (0x0000025C76D6D290 0x0000000000000000 0x000000AB00000001 0x0000025C76D6D290)
    0x00007FF6B84EEDEB (0x000000AB4858A0A0 0x0000000000000005 0x000000AB48580105 0x00007F00B9D43D00)
    0x00007FF6B84EA7B1 (0x000000AB4858A200 0x000000AB4858A300 0x0000025C76DD37A4 0x0000B7E40B3B906E)
    0x00007FF6B84EA334 (0x0000025C76DD37A4 0x0000B7E40B3B93DE 0x000000AB4858A300 0x0000000000000096)
    0x00007FF6B85A2ACC (0x0000025C765F1650 0x00007FF6B9D4595E 0x0000B7E40B3B93BE 0x0000025C7659BA80)
    0x00007FF6B859982C (0x00000000765FD400 0x00007FF6B9D4B5D7 0x0000025C76D39978 0x0000B7E40B3B92EE)
    0x00007FF6B8598C10 (0x0000025C76DD37A0 0x0000000000000000 0x0000025C76D39978 0x0000B7E40B3B95FE)
    0x00007FF6B85A3400 (0x0000025C76DD37A0 0x0000B7E40B3B94CE 0x0000025C765F30E8 0x000000000000003E)
    0x00007FF6B845BB27 (0x0000B7E40B3B944E 0x000000AB4858A7D0 0x0000B7E40B3B947E 0x0000025C765F1650)
    0x00007FF6B845C3B6 (0x0000025C76DD3840 0x0000025C76DD3870 0x0000025C765A1440 0x00007FF6B80B34DD)
    0x00007FF6B82E56CF (0x0000000000000000 0xFFFFFFFFFFFFFFFF 0x0000000000000000 0xFFFFFFFFFFFFFFFF)
    0x00007FF6B82DE54C (0x0000000000000000 0x0000025C765EEA10 0x0000B7E40B3B8C7E 0x0000000000000001)
    0x00007FF6B82E2689 (0x0000000000001837 0x0000025C765EDF18 0x000000000000005A 0x00007FF6B829BD2C)
    0x00007FF6B82E9022 (0x0000000000000000 0x000000AB4858D7F0 0x0000025C765FE570 0x000000AB4858D900)
    0x00007FF6BA34278F (0x0000025C76DD37A0 0x0000000000000000 0x0000025C765FD820 0x0000000000000000)
    0x00007FF6BA340489 (0x000000AB4858D8F8 0x000000AB4858D918 0x0000025C7650E490 0x00007FF6B881FFDA)
    0x00007FF6B94F36C9 (0x0000025C7656E3D0 0x0000B7E40B3BE5CE 0x0000000000000000 0x000000000000000F)
    0x00007FF6B885B562 (0x0000000000000001 0x00007FF600000000 0x00000000000000E8 0x000000AB4858E340)
    0x00007FF6B882250C (0x0000025C76539890 0x000000AB4858D72C 0x0000000000000104 0x000000000000000F)
    0x00007FF6B88C2A2E (0x00000000000003CB 0x0000025C7656D320 0x00000000000003CB 0x00007FFB15D831A2)
    0x00007FF6B65D7164 (0x00007FF6BA361D5C 0x0000000000000000 0x0000000000000000 0x00007FFB135CF24E)
    0x00007FF6B65D4344 (0x0000000000000016 0x0000000020010021 0x000000AB4858E898 0x00007FFB44000145)
    0x00007FF6B65D4066 (0x0000000000000000 0x00007FF6BA35EA31 0x0000000000000000 0x0000000000000000)
    0x00007FF6BA35E9B8 (0x0000000000000000 0x0000000000000000 0x0000000000000000 0x0000000000000000)
    0x00007FFB14C554E0 (0x0000000000000000 0x0000000000000000 0x0000000000000000 0x0000000000000000), BaseThreadInitThunk() + 0x10 bytes(s)
    0x00007FFB15D6485B (0x0000000000000000 0x0000000000000000 0x0000000000000000 0x0000000000000000), RtlUserThreadStart() + 0x2B bytes(s)
    clang: error: clang frontend command failed due to signal (use -v to see invocation)
    clang version 12.0.1
    Target: x86_64-pc-windows-msvc
    Thread model: posix
    InstalledDir: C:\Program Files\LLVM\bin
    clang: note: diagnostic msg:
    ********************

**Comment by AaronBallman** [comment]
Can confirm the issue is present on trunk. It's not just block parameter, it also impacts block return types as well.
```
void foo1(void)
{
    __auto_type blk = ^int(unsigned _BitInt(64) len)
    {
        return 12;
    };
}

void foo2(void)
{
    __auto_type blk = ^unsigned _BitInt(64)(int len)
    {
        return 12;
    };
}
```

**Comment by AaronBallman** [comment]
These issues should now be fixed in b6d9ca14c20f6f982a9fee4bebccf4761400f6aa
